pkg/kubelet/nodestatus/setters.go

```diff
@@ -146,48 +138,72 @@ func NodeAddress(nodeIP net.IP, // typically Kubelet.nodeIP
 				}
 			}
 			node.Status.Addresses = nodeAddresses
-		} else {
-			var ipAddr net.IP
-			var err error
-
-			// 1) Use nodeIP if set
-			// 2) If the user has specified an IP to HostnameOverride, use it
-			// 3) Lookup the IP from node name by DNS and use the first valid IPv4 address.
-			//    If the node does not have a valid IPv4 address, use the first valid IPv6 address.
-			// 4) Try to get the IP from the network interface used as default gateway
+
+			return nil
+		}
+
+		// cloud provider is not set
+		// note this also accounts for the case where --cloud-provider=external
+		// if --cloud-provider=external, we try to set the node address to what we know
+		// at startup and allow the provider to override it later
+		if externalCloudProvider {
 			if nodeIP != nil {
-				ipAddr = nodeIP
-			} else if addr := net.ParseIP(hostname); addr != nil {
-				ipAddr = addr
-			} else {
-				var addrs []net.IP
-				addrs, _ = net.LookupIP(node.Name)
-				for _, addr := range addrs {
-					if err = validateNodeIPFunc(addr); err == nil {
-						if addr.To4() != nil {
-							ipAddr = addr
-							break
-						}
-						if addr.To16() != nil && ipAddr == nil {
-							ipAddr = addr
-						}
-					}
+				if node.ObjectMeta.Annotations == nil {
+					node.ObjectMeta.Annotations = make(map[string]string)
 				}
+				node.ObjectMeta.Annotations[kubeletapis.AnnotationProvidedIPAddr] = nodeIP.String()
+			}
+		}
 
-				if ipAddr == nil {
-					ipAddr, err = utilnet.ChooseHostInterface()
+		// If --cloud-provider=external and node address is already set,
+		// then we return early because provider set addresses should take precedence.
+		// Otherwise, try to look up the node IP and let the cloud provider override it later
+		// This should alleviate a lot of the bootstrapping issues with out-of-tree providers
+		if externalCloudProvider && len(node.Status.Addresses) > 0 {
```

Comment:
We were just in an externalCloudProvider if statement. Why not fold this into the end of that if statement?

Reply:
good catch

Reply:
fixed!